### ISSUE TITLE
Rename `*Reducers.js` files (plural) to `*Reducer.js` (singular).

### ISSUE THREAD

**Issue by jackrzhang** [zulip/zulip-mobile#3016]
While each reducer file does have multiple functions that fit the signature of a 'reducer', which is `(state, action) => state`, each file only exports one function that's utilized elsewhere and invoked in unit tests.

**Comment by jackrzhang** [comment]
@borisyankov Yeah, that's exactly what I've acknowledged in the first comment :) 

To elaborate a bit more - 

Staying with `flagsReducers`; `messageFetchComplete`, `eventNewMessage`, `eventUpdateMessageFlags`, and `markMessagesRead` are specific to *actions*, and I think it makes more sense to think of entire `flagsReducers` module in terms of its specific piece of state.

And that fits with the fact that these files only export a single reducer. So in terms the interface that the module presents, when a new developer or contributor looks at the file for the first time, the singular `reducer` would better communicate what's provided.

**Comment by borisyankov** [comment]
I don't think we ever follow the mantra `the name says what we export'. I think you just invented that to strengthen your case :) For example, the file `reducers.js`, exports a single root reducer, but contains **reducers**.

**Comment by jackrzhang** [comment]
The idea of a module as an interface isn't new.

**Comment by borisyankov** [comment]
Here is my view:
 * calling something `flagsReducer` is correct, as it exports a combined single reducer
 * calling the same thing `flagsReducers` is also correct, as it contains reducers on `flags` state
 * renaming from one to the other is not correct, as it adds complexity to tracking history and non-zero effort for no real benefit

**Comment by gnprice** [comment]
I agree with @jackrzhang that these modules each feel like they're really about the single reducer that it exports, and that maintains a certain subtree of the state. The helper functions like `eventNewMessage` feel like implementation details, and not the point of the module. Which has caused the plurals in the names to not feel right to me.

I'd actually go further -- I don't think those helper functions are properly called "reducers". They don't follow the protocol that Redux expects of a reducer, in a couple of ways:
* Given an action they don't understand, they don't do something reasonable like return the unmodified state -- usually they'll just blow up.
* When passed no state at all, they don't return the desired initial state; they'll start from `undefined`, which neither they nor their selectors accept as a valid state.

Effectively, they count on the exported function at the bottom of the module to handle these essential responsibilities of a Redux reducer for them, and give them a more narrowly focused task. So in fact each of these modules contains just one actual Redux reducer.

---

I don't agree that there's no benefit in fixing these names; as they are, I think the names contribute to confusion about how Redux really works, which is a real cost.

A reason I'd hesitate to go and sweep through renaming them all right now is that I have some other ideas for how we might better organize our Redux code; and as Boris says, there is a cost to renames that affects reading the history, so it's better to move a given chunk of code once than twice.

I spent an afternoon last month prototyping a version of those ideas, but haven't gone back and polished it up. I'll rebase and start a chat discussion with it, rough as it is.

**Comment by jackrzhang** [comment]
> * Given an action they don't understand, they don't do something reasonable like return the unmodified state -- ... 
> * When passed no state at all, they don't return the desired initial state; ...

Yeah, I had also thought of these points; they make clear that the other functions *aren't* Redux reducers.

For reference, the chat discussion is the [#mobile-team stream and topic redux](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/redux).

**Comment by chrisbobbe** [comment]
I’m on mobile now, but I believe the changes described in this PR have been made. @gnprice, it sounds like there are some architectural changes you’re considering in https://github.com/zulip/zulip-mobile/issues/3016#issuecomment-429029878; if they’re still relevant, maybe they could get their own issue and this one could be closed?

**Comment by gnprice** [comment]
Indeed, thanks!

> If the reorganization idea is still relevant, is it being tracked?

Looks like I expanded on that idea [here](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/redux) (thanks to Jack for linking to it above). The prototype branch is here:
https://github.com/zulip/zulip-mobile/compare/master...gnprice:redux ([permalink](https://github.com/zulip/zulip-mobile/compare/master...gnprice:685e7a319))

Some of those ideas, I later completed systematically and merged:
* organizing and documenting the types in our Redux state, now in `src/reduxTypes.js`, circa cc945867a
* inlining the many tiny action handler functions into reducers, in the series around f20c478e5 

The major thing that's still open is:
* For a lot of parts of our state tree, the internal details about that data structure are spread across three places: a `fooReducer.js`, a `fooSelectors.js`, and a `type FooState` in `reduxTypes.js`. I'd like to reorganize so that code that cares about the internal details of `state.foo` is all in one place, perhaps `fooState.js`.

  This can mean less flipping back and forth to understand that code. It also helps make it possible to *encapsulate* those details, i.e. ensure there aren't miscellaneous other bits of code around the app that secretly embed assumptions about that data structure's internals.

There's also a minor thing I see looking back at that branch:
* In preparing the branch, I experimented with using WebStorm, and so it contains a draft WebStorm config.

  My hope was that WebStorm would provide refactoring features, especially for these operations like moving things between files and updating all the imports that refer to them, that in VS Code are broken for Flow codebases. I guess I ended up turning to commands like `perl -i -0pe 's/.../.../g' src/**/*.js` instead; see discussion just today [on `#learning`](https://chat.zulip.org/#narrow/stream/92-learning/topic/perl.20one-liners/near/822352).

  It's still possible WebStorm would be useful to use sometimes. If so, that branch's draft config might be a helpful starting point.
